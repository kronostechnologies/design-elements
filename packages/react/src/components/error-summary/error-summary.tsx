import { MouseEvent, VoidFunctionComponent } from 'react';
import styled from 'styled-components';
import { useTranslation } from '../../i18n/use-translation';
import { StyledLink } from '../route-link/styles/styled-link';
import { DeviceContextProps, useDeviceContext } from '../device-context-provider/device-context-provider';
import { SectionalBanner } from '../sectional-banner/sectional-banner';

type MobileDeviceContext = Pick<DeviceContextProps, 'isMobile'>;
type HeadingTag = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

const Message = styled.p<MobileDeviceContext>`
    font-size: ${(props) => (props.isMobile ? '1rem' : '0.875rem')};
    margin: ${(props) => (props.isMobile ? 'var(--spacing-2x)' : 'var(--spacing-1x)')} 0 0 0;
`;

const ErrorList = styled.ul`
    margin: var(--spacing-2x) 0;
    padding-left: 1rem;
`;

const ErrorItem = styled.li`
    margin-bottom: var(--spacing-1x);
`;

export interface ErrorMessage {
    text: string;
    targetId: string;
}

interface ErrorSummaryProps {
    className?: string;
    /**
     * @default h2
     */
    headingTag?: HeadingTag;
    /**
     * Array of objects with a `text` attribute containing the error message and a `targetId`
     * referencing the target input that will get the focus when clicking the error.
     */
    messages: ErrorMessage[];
}

const handleErrorClick = (targetId: string, event: MouseEvent<HTMLAnchorElement>): void => {
    event.preventDefault();
    document.getElementById(targetId)?.focus();
};

export const ErrorSummary: VoidFunctionComponent<ErrorSummaryProps> = ({
    className,
    headingTag,
    messages,
}) => {
    const { t } = useTranslation('error-summary');
    const { isMobile } = useDeviceContext();

    const errorItems = messages.map((error) => (
        <ErrorItem key={error.targetId}>
            <StyledLink
                $hasLabel={false}
                $isMobile={isMobile}
                href={`#${error.targetId}`}
                onClick={(event) => handleErrorClick(error.targetId, event)}
            >
                {error.text}
            </StyledLink>
        </ErrorItem>
    ));

    return (
        <SectionalBanner
            className={className}
            focusable
            headingTag={headingTag ?? 'h2'}
            title={t('title')}
            type="alert"
        >
            <Message isMobile={isMobile}>{t('message')}</Message>
            <ErrorList>{errorItems}</ErrorList>
        </SectionalBanner>
    );
};
