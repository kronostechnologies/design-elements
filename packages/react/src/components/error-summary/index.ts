export { ErrorSummary } from './error-summary';
export type { ErrorMessage, ErrorSummaryHeadingTag, ErrorSummaryProps } from './error-summary';
