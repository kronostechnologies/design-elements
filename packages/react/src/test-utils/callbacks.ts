export function doNothing(): void {
    // do nothing
}
