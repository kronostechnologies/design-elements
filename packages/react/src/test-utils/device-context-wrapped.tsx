import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
