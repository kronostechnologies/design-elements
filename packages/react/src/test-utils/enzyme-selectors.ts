import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
