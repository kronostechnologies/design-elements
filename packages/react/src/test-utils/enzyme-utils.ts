import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

export async function waitForComponentToPaint<P = {}>(
    wrapper: ReactWrapper<P>,
    amount = 0,
): Promise<void> {
    await act(async () => {
        await new Promise((resolve) => setTimeout(resolve, amount));
        wrapper.update();
    });
}

export function expectFocusToBeOn(element: ReactWrapper): void {
    expect(document.activeElement).toBe(element.getDOMNode());
}
