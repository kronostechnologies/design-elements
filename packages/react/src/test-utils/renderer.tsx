import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
