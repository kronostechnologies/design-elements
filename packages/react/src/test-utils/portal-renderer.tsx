import { render, RenderResult } from '@testing-library/react';
import React, { ReactElement } from 'react';
import { DeviceType } from '../components/device-context-provider/device-context-provider';
import { AllProviders } from './renderer';

export function renderPortalWithProviders(
    component: ReactElement,
    device?: DeviceType,
): RenderResult {
    return render(<AllProviders device={device}>{component}</AllProviders>);
}
