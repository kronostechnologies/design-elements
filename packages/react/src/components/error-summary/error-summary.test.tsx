import { renderWithTheme } from '../../test-utils/renderer';
import { ErrorSummary } from './error-summary';

const errorMessages = [{
    text: 'This is an error',
    targetId: 'error-input',
}, {
    text: 'Another error!',
    targetId: 'error-input-2',
}];

describe('ErrorSummary', () => {
    test('matches the snapshot', () => {
        const tree = renderWithTheme(
            <ErrorSummary messages={errorMessages} />,
        );

        expect(tree).toMatchSnapshot();
    });
});
