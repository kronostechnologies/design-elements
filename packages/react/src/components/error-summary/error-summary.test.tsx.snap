// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorSummary matches the snapshot 1`] = `
.c9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #006296;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 0.875rem;
  line-height: 1.5rem;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c9:hover {
  color: #003A5A;
}

.c9:visited {
  color: #62a;
}

.c9:focus {
  outline: none;
}

.c9:focus {
  outline: none;
  box-shadow: 0 0 0 2px #84C6EA;
  box-shadow: 0 0 0 1px #006296,0 0 0 3px #84C6EA;
}

.c9:focus:not(:focus-visible) {
  box-shadow: none;
}

.c9:focus-visible {
  outline: none;
  box-shadow: 0 0 0 2px #84C6EA;
  box-shadow: 0 0 0 1px #006296,0 0 0 3px #84C6EA;
}

.c2 {
  grid-area: icon;
  margin-top: 2px;
}

.c0 {
  background-color: color-alert-02;
  border: 1px solid #CD2C23;
  box-sizing: border-box;
  line-height: 24px;
  padding: var(--spacing-2x);
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0:focus {
  outline: none;
}

.c0:focus {
  outline: none;
  border-color: #006296;
  box-shadow: 0 0 0 2px #84C6EA;
}

.c0:focus:not(:focus-visible) {
  border-color: inherit;
  box-shadow: none;
}

.c0 .c1 {
  color: #CD2C23;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  height: 1rem;
  width: 1rem;
}

.c5 {
  font-size: 0.875rem;
  margin: var(--spacing-1x) 0 0 0;
}

.c3 {
  box-sizing: border-box;
  grid-area: content;
  padding-left: var(--spacing-2x);
}

.c4 {
  font-size: 1rem;
  font-weight: var(--font-semi-bold);
  margin: 0;
}

.c6 {
  font-size: 0.875rem;
  margin: var(--spacing-1x) 0 0 0;
}

.c7 {
  margin: var(--spacing-2x) 0;
  padding-left: 1rem;
}

.c8 {
  margin-bottom: var(--spacing-1x);
}

<section
  aria-atomic="true"
  aria-labelledby="uuid1"
  aria-live="assertive"
  class="c0"
  role="alert"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="c1 c2"
    color="currentColor"
    focusable="false"
    height="20"
    width="20"
  />
  <div
    class="c3"
  >
    <h2
      class="c4"
      id="uuid1"
    >
      The form contains one or more errors
    </h2>
    <div
      class="c5"
    >
      <p
        class="c6"
      >
        Please correct them to proceed to the next step.
      </p>
      <ul
        class="c7"
      >
        <li
          class="c8"
        >
          <a
            class="c9"
            href="#error-input"
          >
            This is an error
          </a>
        </li>
        <li
          class="c8"
        >
          <a
            class="c9"
            href="#error-input-2"
          >
            Another error!
          </a>
        </li>
      </ul>
    </div>
  </div>
</section>
`;
